torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp

```diff
@@ -0,0 +1,194 @@
+#include <torch/csrc/jit/codegen/cuda/arith.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
+
+#include <torch/csrc/jit/codegen/cuda/lower_thread_predicate.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+
+const static std::unordered_map<ParallelType, int> pt_to_offset{
+    {ParallelType::BIDx, 0},
+    {ParallelType::BIDy, 1},
+    {ParallelType::BIDz, 2},
+    {ParallelType::TIDx, 3},
+    {ParallelType::TIDy, 4},
+    {ParallelType::TIDz, 5}};
+
+const static std::unordered_map<int, ParallelType> offset_to_pt{
+    {0, ParallelType::BIDx},
+    {1, ParallelType::BIDy},
+    {2, ParallelType::BIDz},
+    {3, ParallelType::TIDx},
+    {4, ParallelType::TIDy},
+    {5, ParallelType::TIDz}};
+
+static constexpr int num_p_type = 6;
+
+namespace {
+
+void flip_true(std::bitset<num_p_type>& bits, const ParallelType p_type) {
+  if (pt_to_offset.find(p_type) == pt_to_offset.end()) {
+    TORCH_INTERNAL_ASSERT(false, "Could not recognize parallel type.");
+  }
+  bits[pt_to_offset.at(p_type)] = true;
+}
+
+Val* threadPredicate(int i) {
+  if (offset_to_pt.find(i) == offset_to_pt.end()) {
+    TORCH_INTERNAL_ASSERT(
+        false,
+        "Invalid int for predicate computation, should be from [0-5], but recieved, ",
+        i,
+        ".");
+  }
+  return eq(
+      new NamedScalar(stringifyThread(offset_to_pt.at(i)), DataType::Int),
+      new Int(0));
+}
+
+Bool* getThreadPredicate(std::bitset<num_p_type> bits) {
+  if (bits.none())
+    return new Bool(true);
+
+  Val* pred = nullptr;
+
+  for (int i = 0; i < num_p_type; i++) {
+    if (bits[i]) {
+      if (pred == nullptr) {
+        pred = threadPredicate(i);
+      } else {
+        pred = andOp(pred, threadPredicate(i));
+      }
+    }
+  }
+
+  // Should never be hit.
+  TORCH_INTERNAL_ASSERT(pred != nullptr);
+
+  TORCH_INTERNAL_ASSERT(
+      pred->getDataType().value() == DataType::Bool,
+      "Tried to return a predicate that is not a bool val.");
+
+  return pred->as<Bool>();
+}
+
+} // namespace
+
+std::bitset<num_p_type> ThreadPredicates::getThreadPredicates(
+    const TensorView* tv) {
+  TORCH_INTERNAL_ASSERT(
+      thread_predicates.find(tv) != thread_predicates.end(),
```

Comment:
as far as I can tell, TORCH_CHECK and TORCH_INTERNAL_ASSERTS are present in all build flavors (build/release) - this is a very good thing (TM), but I think we should be mindful of duplicating potentially expensive computation/lookups in checks - am I missing anything?